site/_posts/2021-09-16-how-to-authenticate-your-react-app.md

```diff
@@ -0,0 +1,1087 @@
+
+app.use(cors({
+  origin: true,
+  credentials: true
+}));
+
+// use routes
+app.use('/user', userRoute);
+app.use('/login', loginRoute);
+
+// start server
+app.listen(config.serverPort, () => console.log(`FusionAuth example app listening on port ${config.serverPort}.`));
+
+```
+
+When a user goes to `/login`, we redirect them from the Express sever to FusionAuth’s login page. One of the main reasons we installed FusionAuth is so that we don’t have to handle the login process at all!  
+
+*Note: there is a way to access the FusionAuth pages in an iframe so that the user doesn't have to be redirected, but that is outside the scope of this tutorial.*
+
+Now, lets configure our `server/routes/login.js` file to do just that:
+
+```js
+const express = require('express');
+const router = express.Router();
+const config = require('../../client/src/config');
+
+router.get('/', (req, res) => {
+  res.redirect(`http://localhost:${config.fusionAuthPort}/oauth2/authorize?client_id=${config.clientID}&redirect_uri=${config.redirectURI}&response_type=code`);
+});
+
+module.exports = router;
+```
+
+If that URI looks a bit messy, it’s because of the additional query parameters, which FusionAuth needs to process our request:
+
+- `client_id` tells FusionAuth which app is making the request
+- `redirect_uri` tells FusionAuth where to redirect the user to after login
+- `response_type` tells FusionAuth which OAuth grant we’re using (Authorization Code in this example)
+
+This is all part of the standard OAuth Authorization Code grant.
+
+Try navigating to `localhost:3001/login`. If you see a FusionAuth login form, you're all set!
+
+{% include _image.liquid src="/assets/img/blogs/fusionauth-example-react-2021/login.png" alt="login screen" class="img-fluid" figure=false %}
+
+When you successfully authenticate, you’ll just see `Cannot GET /oauth-callback`, because the `/oauth-callback` route doesn’t exist yet. Remember you added `localhost:3001/oauth-callback` as an "Authorized redirect URL" in the FusionAuth admin panel, *and* as our `redirectURI` in `config.js`. This is the location where FusionAuth redirects the browser back to after authentication in order to complete the OAuth grant.   
+
+But why specify this? Because without specifying where to send the app after authenticating, a bad actor could put in their own redirect URL, sending the browser to their malicious server and gaining access to a token that could be used to view resources as a user. That would be no good!
+
+### Exchange the Authorization Code for an Access Token
+
+An Authorization Code isn’t enough to access the user’s resources, though. For that, we need an Access Token. This is standard OAuth, not something unique to FusionAuth. This step is called the Code Exchange, because we send the auth code to FusionAuth’s `/token` endpoint and receive an Access Token in exchange.  Then that Access Token is passed to the resource server in exchange for the desired resources.  
```

Comment:
```suggestion
If you look at the URL when you see the `Cannot GET /oauth-callback` message, you'll see a parameter named `code`. This is the Authorization Code, which is the namesake of this grant. However, an Authorization Code isn’t enough to access the user’s resources, though. For that, we need an Access Token. This is standard OAuth, not something unique to FusionAuth. This step is called the Code Exchange, because we send the auth code to FusionAuth’s `/token` endpoint and receive an Access Token in exchange.  Then that Access Token is passed to the resource server in exchange for the desired resources.  
```